future/interactions.go

```diff
@@ -0,0 +1,166 @@
+package future
+
+import (
+	"crypto/sha512"
+	"errors"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/algorand/go-algorand-sdk/abi"
+)
+
+type Arg struct {
+	Name    string `json:"name"`
+	AbiType string `json:"type"`
+	Desc    string `json:"desc,omitempty"`
+}
+
+type Return struct {
+	AbiType string `json:"type"`
+	Desc    string `json:"desc,omitempty"`
+}
+
+type Method struct {
+	Name    string `json:"name"`
+	Desc    string `json:"desc,omitempty"`
+	Args    []Arg  `json:"args"`
+	Returns Return `json:"returns"`
+}
+
+func MethodFromSignature(methodStr string) (Method, error) {
+	openCnt := strings.Count(methodStr, "(")
+	if openCnt == 0 {
+		return Method{}, errors.New("Method signature has invalid format")
+	}
+
+	closeCnt := strings.Count(methodStr, ")")
+	if closeCnt != openCnt {
+		return Method{}, errors.New("Method signature has invalid format")
+	}
+
+	openIdx := strings.Index(methodStr, "(")
+	parenCnt := 0
+	closeIdx := -1
+	for pos := openIdx; pos < len(methodStr); pos++ {
+		if methodStr[pos] == '(' {
+			parenCnt++
+		} else if methodStr[pos] == ')' {
+			parenCnt--
+		}
+
+		if parenCnt == 0 {
+			closeIdx = pos
+			break
+		}
+	}
+
+	if closeIdx == -1 {
+		return Method{}, errors.New("Method signature has invalid format")
+	}
+
+	// signature must include method name
+	if openIdx == 0 {
+		return Method{}, errors.New("Method signature has invalid format")
+	}
+
+	name := methodStr[:openIdx]
+	namePattern := "[_a-zA-Z][_a-zA-Z0-9]+"
+	_, err := regexp.MatchString(namePattern, name)
```

Comment:
nit: Pavel taught me before that you can put this in `init` and precompile the regex, which can save some constant in perf (also we might do not want `_` in leading symbol)